src/JuMP.jl

```diff
@@ -53,38 +53,34 @@ include("utils.jl")
 const _MOIVAR = MOI.VariableIndex
 const _MOICON{F,S} = MOI.ConstraintIndex{F,S}
 
-# OptimizerFactory was deprecated in JuMP 0.21 and should be removed when
-# with_optimizer is removed.
-struct OptimizerFactory
-    # The constructor can be
-    # * `Function`: a function, or
-    # * `DataType`: a type, or
-    # * `UnionAll`: a type with missing parameters.
-    constructor
-    args::Tuple
-    kwargs # type changes from Julia v0.6 to v0.7 so we leave it untyped for now
-end
-
-function with_optimizer(constructor,
-                        args...; kwargs...)
+function with_optimizer(constructor; kwargs...)
+    if isempty(kwargs)
+        deprecation_message = """
+with_optimizer is deprecated, replace `with_optimizer(Ipopt.Optimizer)` by `Ipopt.Optimizer`.
+"""
+        Base.depwarn(deprecation_message, :with_optimizer)
+        return constructor
+    else
+        deprecation_message = """
+`with_optimizer` is deprecated, replace `with_optimizer(Ipopt.Optimizer, max_cpu_time=60.0)`
+by `MOI.OptimizerWithAttributes(Ipopt.Optimizer, "max_cpu_time" => 60.0)`.
```

Comment:
Let's re-export OptimizerWithAttributes. This is essential and shouldn't be hidden in a submodule.

Reply:
From the JuMP-dev called, it seems it should be `JuMP.optimizer_with_attributes`

Reply:
That works for me also.